### PR TITLE
Centralize argument parsing

### PR DESCRIPTION
**Pull request by tgharib**
Solves https://github.com/GrayHatter/uTox/issues/142 and https://github.com/notsecure/uTox/issues/1060

### CODE REVIEW

File: src/android/main.c [hunk 1]
```diff
@@ -519,6 +519,24 @@ void config_osdefaults(UTOX_SAVE *r)
...
+    if (launch_at_startup == 1 || launch_at_startup == -1) {
+        debug("Start on boot not supported on this OS!\n");
+    }
+    
+    if (set_show_window == 1 || set_show_window == -1) {
+        debug("Showing/hiding windows not supported on this OS!\n");
+    }
+    
+    if (no_updater == 1) {
+        debug("Disabling the updater is not supported on this OS. Updates are managed by the play store.\n");
```
Comment (cebe) [inline src/android/main.c]: not necessarily Play store but "App Store" in general. e.g. fdroid.

File: src/main.c [hunk 2]
```diff
@@ -1,4 +1,104 @@
...
+                } else if (!strcmp(optarg, "dark")) {
+                    theme = THEME_DARK;
+                } else if (!strcmp(optarg, "light")) {
+                    theme = THEME_LIGHT;
+                } else if (!strcmp(optarg, "highcontrast")) {
+                    theme = THEME_HIGHCONTRAST;
+                } else if (!strcmp(optarg, "zenburn")) {
+                    theme = THEME_ZENBURN;
+                } else {
+                    debug("Please specify correct theme (please check user manual for list of correct values).");
```
Comment (cebe) [inline src/main.c]: could just list valid values here as we have a complete list above anyway.

File: src/main.c [hunk 3]
```diff
@@ -1,4 +1,104 @@
...
+    
+    static struct option long_options[] = {
+        {"theme",         required_argument,   NULL,  't'},
+        {"portable",      no_argument,            NULL,  'p'},
+        {"set",              required_argument,   NULL,  's'},
+        {"unset",          required_argument,   NULL,  'u'},
+        {"no-updater", no_argument,            NULL,  'n'},
+        {"version",       no_argument,            NULL,  'v'},
+        {"help",            no_argument,            NULL,  'h'},
+        {0, 0, 0, 0}
```
Comment (benwaffle) [inline src/main.c]: AHHHHHH they don't line up

File: src/main.c [hunk 4]
```diff
@@ -1,4 +1,104 @@
...
+                break;
+                
+            case 'u':
+                if (!strcmp(optarg, "start-on-boot")) {
+                    *launch_at_startup = -1;
+                } else {
+                    debug("Please specify a correct unset option (please check user manual for list of correct values).");
+                    exit(EXIT_FAILURE);
+                }
+                break;
```
Comment (benwaffle) [inline src/main.c]: I think we should remove set & unset. 1) they only handle 3 (or 1) flags and 2) i doubt anybody ever used them
If we were to keep it, it should be able to set & unset any config value

File: src/main.c [hunk 5]
```diff
@@ -1,4 +1,104 @@
...
+            case 'h':
+                printf("µTox - Lightweight Tox client version %s.\n\n", VERSION);
+                printf("The following options are available:\n\n");
+                printf("  -t --theme=<theme-name>  Specify a UI theme, where <theme-name> can be one of default, dark, light, highcontrast, zenburn.\n");
+                printf("  -p --portable            Launch in portable mode: All data will be saved to the tox folder in the current working directory.\n");
+                printf("  -s --set=<option>        Set an option: start-on-boot, show-window, hide-window.\n");
+                printf("  -u --unset=<option>      Unset an option: start-on-boot.\n");
+                printf("  -n --no-updater          Disable the updater.\n");
+                printf("  -v --version             Print the version and exit.\n");
+                printf("  -h --help                Shows this help text.\n\n");
```
Comment (benwaffle) [inline src/main.c]: `puts()` is a tiny bit cleaner

**Comment by cebe** [comment]
travis fails on Mac, btw: https://travis-ci.org/GrayHatter/uTox/jobs/99537541#L2421

File: src/main.c (hunk 2 above)
Comment (tgharib) [inline src/main.c]: @cebe Not sure what you mean here.

File: src/main.c (hunk 3 above)
Comment (tgharib) [inline src/main.c]: @benwaffle I swear they lined up in NetBeans! NetBeans lied to me. :(

File: src/main.c (hunk 2 above)
Comment (cebe) [inline src/main.c]: instead of "please check user manual for list of correct values", say: "valid values are ..."

File: src/main.c (hunk 4 above)
Comment (tgharib) [inline src/main.c]: I agree. Do you think we should just have four flags: `enable-start-on-boot`, `disable-start-on-boot`, `show-window` and `hide-window`?

**Comment by stal888** [comment]
Is there a reason you don't use bools for launch_at_startup/set_show_window?

File: src/main.c (hunk 2 above)
Comment (GrayHatter) [inline src/main.c]: I'm undecided, I like @cebe's suggestion, but that's just one more place we have to do upkeep...

Also, the manual also tells you how to make a custom one, I'd kinda like to push people that way

**Comment by GrayHatter** [comment]
@stal888 because it's a set, unset. We don't want to change them without that flag so we can't just if() it, else we'll unset when it's null. (I already asked that question)

File: src/main.c (hunk 4 above)
Comment (benwaffle) [inline src/main.c]: i don't think any of those options are useful

Comment (GrayHatter) [inline src/main.c]: Says the Linux user... Those are windows settings.

Comment (GrayHatter) [inline src/main.c]: Also, if they tray doesn't draw, and uTox is set to start in tray... You're kinda fucked

Comment (benwaffle) [inline src/main.c]: those settings can be set from the GUI
and if they tray doesn't draw you have a bug to fix

**Comment by tsudoko** [comment]
Why are you using camel case in function names? Other parts of codebase use snake case.

**Comment by tgharib** [comment]
Alright, I've implemented your suggestions and the pull request should be complete. It needs to be tested. I can't seem to compile for windows so I've only tested on linux.

**Comment by GrayHatter** [comment]
@tgharib 
```
  CC    src/windows/main.o
src/windows/main.c: In function ‘CommandLineToArgvA’:
src/windows/main.c:1193:5: warning: suggest parentheses around assignment used as truth value [-Wparentheses]
     while (a = CmdLine[i]) {
     ^
src/windows/main.c: In function ‘WinMain’:
src/windows/main.c:1346:15: error: ‘arglist’ undeclared (first use in this function)
     LocalFree(arglist);
               ^
src/windows/main.c:1346:15: note: each undeclared identifier is reported only once for each function it appears in
src/windows/main.c:1426:5: warning: implicit declaration of function ‘dnd_init’ [-Wimplicit-function-declaration]
     dnd_init(hwnd);
     ^
Makefile:134: recipe for target 'src/windows/main.o' failed
make: *** [src/windows/main.o] Error 1
```

**Comment by tgharib** [comment]
Ping. Hopefully no more compilation errors.

**Comment by GrayHatter** [comment]
testing this today

File: src/main.c [hunk 6]
```diff
@@ -1,4 +1,104 @@
...
+                    debug("Please specify a correct set option (please check user manual for list of correct values).");
+                    exit(EXIT_FAILURE);
+                }
+                break;
+                
+            case 'u':
+                if (!strcmp(optarg, "start-on-boot")) {
+                    *should_launch_at_startup = -1;
+                } else {
+                    debug("Please specify a correct unset option (please check user manual for list of correct values).");
```
Comment (tsudoko) [inline src/main.c]: needs `\n`

File: src/main.c [hunk 7]
```diff
@@ -1,4 +1,104 @@
...
+                
+            case 's':
+                if (!strcmp(optarg, "start-on-boot")) {
+                    *should_launch_at_startup = 1;
+                } else if (!strcmp(optarg, "show-window")) {
+                    *set_show_window = 1;
+                } else if (!strcmp(optarg, "hide-window")) {
+                    *set_show_window = -1;
+                } else {
+                    debug("Please specify a correct set option (please check user manual for list of correct values).");
```
Comment (tsudoko) [inline src/main.c]: needs `\n`

File: src/main.c (hunk 2 above)
Comment (tsudoko) [inline src/main.c]: needs `\n`

**Comment by GrayHatter** [comment]
themes now work on windows in release-v0.6

**Comment by tgharib** [comment]
Any reason why this has been closed? I couldn't test on Windows due to outdated documentation at the time so if that's the issue, I can take another crack at compiling on Windows.

**Comment by GrayHatter** [comment]
Closed because it's already staged in the branch release-v0.6 and will get
pushed to master tonight

**Pull request closed without merging**